core/txpool/legacypool/list.go

```diff
@@ -286,6 +286,12 @@ type list struct {
 	costcap   *uint256.Int // Price of the highest costing transaction (reset only if exceeds balance)
 	gascap    uint64       // Gas limit of the highest spending transaction (reset only if exceeds block limit)
 	totalcost *uint256.Int // Total cost of all transactions in the list
+
+	// OP-Stack adds a backpointer to the pool's rollup cost function.
+	// A list always belongs to a fixed pool, so it's ok to use a reference instead of
+	// always passing the rollup cost function as an argument to every function.
+	// It should not be accessed directly, but through the rollupCostFn method.
+	rollupCostFnPtr *txpool.RollupCostFunc
```

Comment:
```suggestion
	rollupCostFn txpool.RollupCostFunc
```
Couldn't we do this and avoid the indirection below?

Reply:
I don't think so because the function is set anew in every `LegacyPool.reset()` call:
```go
	if costFn := types.NewTotalRollupCostFunc(pool.chainconfig, statedb); costFn != nil {
		pool.rollupCostFn = func(tx types.RollupTransaction) *uint256.Int {
			return costFn(tx, newHead.Time)
		}
	}
```
that's the reason I use a pointer, so that it always points at the right function because the function is stored at this field.

Reply:
I tried this out here and all the tests still passed https://github.com/ethereum-optimism/op-geth/compare/seb/txpool-rollup-cost...gk/expmt?expand=1 . So I guess this means either it's an OK change or we don't have great test coverage?

Reply:
The coverage for a changing chain head is indeed not there. Worth adding a test for this 👍🏻 

I had another idea how to implement this better. Currently, the rollup cost function is re-initialized every time in `LegacyPool.reset()`. But the pool also stores the `currentHead`. So we can actually already initialize the rollup cost function in the constructor and let it read the block number from the `currentHead`. Then we can just set it in the `list`s and also don't need to reinitialize for every new head in `reset`. I will do this.

Reply:
+1 on making this change

Reply:
Actually, a new stateDB also is needed for every new time the function is initialized. And while that is tracked in the pool, it means you need a reference to the pool.

Instead, I think adding a method to the pool is cleaner: the method-receiver is then bound to the pool, and the function can be passed down into the list. Meanwhile, the reset method can create the inner rollup-cost function for every new block (and enjoy the caching etc. of parsed cost parameters), and the pool method can then just call that.
No explicit pointers needed that way.

Reply:
Yeah I also saw that the statedb is generated anew on Init and every reset...we would need the rollup cost functions to accept a statedb provider instead of a mere pointer for this to work. Replaced the function pointer by an interface that provides the RollupCostFunc. This is probably compiled to almost the same code, but it's nicer than passing bare function pointers around.

I've also added the reset of the cost func to the pool's `Init()` function, so it's available before any call to `reset()`.

Reply:
Added a test of changing rollup parameters to confirm this is working as intended.